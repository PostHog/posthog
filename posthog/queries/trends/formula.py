import math
from itertools import accumulate
from typing import Any, Dict, List

from posthog.clickhouse.kafka_engine import trim_quotes_expr
from posthog.client import sync_execute
from posthog.constants import NON_TIME_SERIES_DISPLAY_TYPES, TRENDS_CUMULATIVE
from posthog.models.filters.filter import Filter
from posthog.models.team import Team
from posthog.queries.breakdown_props import get_breakdown_cohort_name
from posthog.queries.trends.util import parse_response


class TrendsFormula:
    def _run_formula_query(self, filter: Filter, team: Team):
        letters = [chr(65 + i) for i in range(0, len(filter.entities))]
        queries = []
        params: Dict[str, Any] = {}
        for idx, entity in enumerate(filter.entities):
            sql, entity_params, _ = self._get_sql_for_entity(filter, team, entity)  # type: ignore
            sql = sql.replace("%(", f"%({idx}_")
            entity_params = {f"{idx}_{key}": value for key, value in entity_params.items()}
            queries.append(sql)
            params = {**params, **entity_params}

        breakdown_value = (
            ", sub_A.breakdown_value"
            if filter.breakdown_type == "cohort"
            else f", {trim_quotes_expr('sub_A.breakdown_value')}"
        )
        is_aggregate = filter.display in NON_TIME_SERIES_DISPLAY_TYPES

        sql = """SELECT
            {date_select}
            arrayMap(({letters_select}) -> {formula}, {selects})
            {breakdown_value}
            {max_length}
            FROM ({first_query}) as sub_A
            {queries}
        """.format(
            date_select="'' as date," if is_aggregate else "sub_A.date,",
            letters_select=", ".join(letters),
            formula=filter.formula,  # formula is properly escaped in the filter
            # Need to wrap aggregates in arrays so we can still use arrayMap
            selects=", ".join(
                [
                    (f"[sub_{letter}.data]" if is_aggregate else f"arrayResize(sub_{letter}.data, max_length, 0)")
                    for letter in letters
                ]
            ),
            breakdown_value=breakdown_value if filter.breakdown else "",
            max_length=""
            if is_aggregate
            else ", arrayMax([{}]) as max_length".format(", ".join(f"length(sub_{letter}.data)" for letter in letters)),
            first_query=queries[0],
            queries="".join(
                [
                    "FULL OUTER JOIN ({query}) as sub_{letter} ON sub_A.breakdown_value = sub_{letter}.breakdown_value ".format(
                        query=query, letter=letters[i + 1]
                    )
                    for i, query in enumerate(queries[1:])
                ]
            )
            if filter.breakdown
            else "".join(
                [" CROSS JOIN ({}) as sub_{}".format(query, letters[i + 1]) for i, query in enumerate(queries[1:])]
            ),
        )
        result = sync_execute(sql, params)
        response = []
        for item in result:
            additional_values: Dict[str, Any] = {
                "label": self._label(filter, item),
            }
            if is_aggregate:
                additional_values["data"] = []
                additional_values["aggregated_value"] = item[1][0]
            else:
                additional_values["data"] = [
                    round(number, 2) if not math.isnan(number) and not math.isinf(number) else 0.0 for number in item[1]
                ]
                if filter.display == TRENDS_CUMULATIVE:
                    additional_values["data"] = list(accumulate(additional_values["data"]))
            additional_values["count"] = float(sum(additional_values["data"]))
            response.append(parse_response(item, filter, additional_values=additional_values))
        return response

    def _label(self, filter: Filter, item: List) -> str:
        if filter.breakdown:
            if filter.breakdown_type == "cohort":
                return get_breakdown_cohort_name(item[2])
            return item[2]
        return "Formula ({})".format(filter.formula)
