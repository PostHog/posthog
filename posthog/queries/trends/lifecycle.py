from datetime import datetime
from typing import Callable, Dict, List, Tuple

from django.db.models.query import Prefetch
from rest_framework.request import Request

from posthog.client import sync_execute
from posthog.models.entity import Entity
from posthog.models.entity.util import get_entity_filtering_params
from posthog.models.filters import Filter
from posthog.models.filters.mixins.utils import cached_property
from posthog.models.person.util import get_persons_by_uuids
from posthog.models.team import Team
from posthog.queries.event_query import EventQuery
from posthog.queries.person_query import PersonQuery
from posthog.queries.trends.sql import LIFECYCLE_PEOPLE_SQL, LIFECYCLE_SQL
from posthog.queries.trends.util import parse_response
from posthog.queries.util import parse_timestamps

# Lifecycle takes an event/action, time range, interval and for every period, splits the users who did the action into 4:
#
# 1. NEW - Users who did the action during interval and were also created during that period
# 2. RESURRECTING - Users who did the action during this interval, but not one prior
# 3. RETURNING - Users who did the action during this interval and prior one
# 4. DORMANT - Users who did not do the action during this period but did an action the previous period
#
# To do this, we need for every period (+1 prior to the first period), list of person_ids who did the event/action
# during that period and their creation dates.


class Lifecycle:
    def _format_lifecycle_query(self, entity: Entity, filter: Filter, team: Team) -> Tuple[str, Dict, Callable]:
        event_query, event_params = LifecycleEventQuery(team=team, filter=filter).get_query()

        return (
            LIFECYCLE_SQL.format(events_query=event_query, interval_expr=filter.interval),
            event_params,
            self._parse_result(filter, entity, team),
        )

    def _parse_result(self, filter: Filter, entity: Entity, team: Team) -> Callable:
        def _parse(result: List) -> List:
            res = []
            for val in result:
                label = "{} - {}".format(entity.name, val[2])
                additional_values = {"label": label, "status": val[2]}
                parsed_result = parse_response(val, filter, additional_values=additional_values)
                res.append(parsed_result)

            return res

        return _parse

    def get_people(
        self,
        filter: Filter,
        team: Team,
        target_date: datetime,
        lifecycle_type: str,
        request: Request,
        limit: int = 100,
    ):
        event_query, event_params = LifecycleEventQuery(team=team, filter=filter).get_query()

        result = sync_execute(
            LIFECYCLE_PEOPLE_SQL.format(events_query=event_query, interval_expr=filter.interval),
            {
                **event_params,
                "status": lifecycle_type,
                "target_date": target_date,
                "offset": filter.offset,
                "limit": limit,
            },
        )
        people = get_persons_by_uuids(team=team, uuids=[p[0] for p in result])
        people = people.prefetch_related(Prefetch("persondistinctid_set", to_attr="distinct_ids_cache"))

        from posthog.api.person import PersonSerializer

        return PersonSerializer(people, many=True).data


class LifecycleEventQuery(EventQuery):
    _filter: Filter

    def get_query(self):
        date_query, date_params = self._get_date_filter()
        self.params.update(date_params)

        prop_query, prop_params = self._get_prop_groups(
            self._filter.property_groups, person_id_joined_alias=f"{self.DISTINCT_ID_TABLE_ALIAS}.person_id"
        )

        self.params.update(prop_params)

        person_query, person_params = self._get_person_query()
        self.params.update(person_params)

        groups_query, groups_params = self._get_groups_query()
        self.params.update(groups_params)

        entity_params, entity_format_params = get_entity_filtering_params(
            entity=self._filter.entities[0], team_id=self._team_id, table_name=self.EVENT_TABLE_ALIAS
        )
        self.params.update(entity_params)

        return (
            f"""
            SELECT DISTINCT
                {self.DISTINCT_ID_TABLE_ALIAS}.person_id as person_id,
                dateTrunc(%(interval)s, toDateTime(events.timestamp, %(timezone)s)) AS period,
                toDateTime(person.created_at, %(timezone)s) AS created_at
            FROM events AS {self.EVENT_TABLE_ALIAS}
            {self._get_distinct_id_query()}
            {person_query}
            {groups_query}
            WHERE team_id = %(team_id)s
            {entity_format_params["entity_query"]}
            {date_query}
            {prop_query}
        """,
            self.params,
        )

    @cached_property
    def _person_query(self):
        return PersonQuery(self._filter, self._team_id, self._column_optimizer, extra_fields=["created_at"],)

    def _get_date_filter(self):
        _, _, date_params = parse_timestamps(filter=self._filter, team=self._team)
        params = {**date_params, "interval": self._filter.interval}
        # :TRICKY: We fetch all data even for the period before the graph starts up until the end of the last period
        return (
            f"""
            AND timestamp >= toDateTime(dateTrunc(%(interval)s, toDateTime(%(date_from)s))) - INTERVAL 1 {self._filter.interval}
            AND timestamp < toDateTime(dateTrunc(%(interval)s, toDateTime(%(date_to)s))) + INTERVAL 1 {self._filter.interval}
        """,
            params,
        )

    def _determine_should_join_distinct_ids(self) -> None:
        self._should_join_distinct_ids = True

    def _determine_should_join_persons(self) -> None:
        self._should_join_persons = True
