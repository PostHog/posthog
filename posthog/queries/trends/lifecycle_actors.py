from typing import Dict, Optional, Tuple

from posthog.queries.actor_base_query import ActorBaseQuery
from posthog.queries.trends.lifecycle import LifecycleEventQuery
from posthog.models.filters.lifecycle_filter import LifecycleFilter

from posthog.queries.trends.sql import LIFECYCLE_PEOPLE_SQL


class LifecycleActors(ActorBaseQuery):
    event_query_class = LifecycleEventQuery
    _filter: LifecycleFilter

    QUERY_TYPE = "lifecycle"

    def actor_query(self, limit_actors: Optional[bool] = True) -> Tuple[str, Dict]:
        events_query, event_params = self.event_query_class(
            filter=self._filter,
            team=self._team,
            person_on_events_mode=self._team.person_on_events_mode,
        ).get_query()

        lifecycle_type = self._filter.lifecycle_type
        target_date = self._filter.target_date

        return (
            LIFECYCLE_PEOPLE_SQL.format(
                events_query=events_query,
                limit=f'{"LIMIT %(limit)s" if limit_actors else ""}',
                offset=f'{"OFFSET %(offset)s" if limit_actors else ""}',
            ),
            {
                **event_params,
                **self._filter.hogql_context.values,
                "offset": self._filter.offset,
                "limit": self._filter.limit,
                "status": lifecycle_type,
                "target_date": target_date,
            },
        )
