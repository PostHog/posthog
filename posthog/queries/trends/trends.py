import copy
import threading
from datetime import datetime, timedelta
from itertools import accumulate
from typing import (
    Any,
    Callable,
    Dict,
    List,
    Optional,
    Tuple,
    Union,
    cast,
)

import pytz
from dateutil import parser
from django.db.models.query import Prefetch

from posthog.client import sync_execute
from posthog.constants import TREND_FILTER_TYPE_ACTIONS, TRENDS_CUMULATIVE, TRENDS_LIFECYCLE, TRENDS_LINEAR
from posthog.models.action import Action
from posthog.models.action_step import ActionStep
from posthog.models.entity import Entity
from posthog.models.filters import Filter
from posthog.models.team import Team
from posthog.queries.base import handle_compare
from posthog.queries.trends.breakdown import TrendsBreakdown
from posthog.queries.trends.formula import TrendsFormula
from posthog.queries.trends.lifecycle import Lifecycle
from posthog.queries.trends.total_volume import TrendsTotalVolume
from posthog.utils import generate_cache_key, get_safe_cache


class Trends(TrendsTotalVolume, Lifecycle, TrendsFormula):
    def _get_sql_for_entity(self, filter: Filter, team: Team, entity: Entity) -> Tuple[str, Dict, Callable]:
        if filter.breakdown:
            sql, params, parse_function = TrendsBreakdown(
                entity, filter, team, using_person_on_events=team.actor_on_events_querying_enabled
            ).get_query()
        elif filter.shown_as == TRENDS_LIFECYCLE:
            sql, params, parse_function = self._format_lifecycle_query(entity, filter, team)
        else:
            sql, params, parse_function = self._total_volume_query(entity, filter, team)

        return sql, params, parse_function

    # Use cached result even on refresh if team has strict caching enabled
    def get_cached_result(self, filter: Filter, team: Team) -> Optional[List[Dict[str, Any]]]:

        if not team.strict_caching_enabled or filter.breakdown or filter.display != TRENDS_LINEAR:
            return None

        cache_key = generate_cache_key(f"{filter.toJSON()}_{team.pk}")
        cached_result_package = get_safe_cache(cache_key)
        cached_result = (
            cached_result_package.get("result")
            if cached_result_package and isinstance(cached_result_package, dict)
            else None
        )

        if not cached_result:
            return None

        _is_present = self.is_present_timerange(cached_result, filter, team)

        return cached_result if _is_present else None

    # Determine if the current timerange is present in the cache
    def is_present_timerange(self, cached_result: List[Dict[str, Any]], filter: Filter, team: Team) -> bool:
        if (
            len(cached_result) > 0
            and cached_result[0].get("days")
            and cached_result[0].get("data")
            and len(cached_result[0]["days"]) > 0
            and len(cached_result[0]["days"]) == len(cached_result[0]["data"])
        ):
            latest_date = cached_result[0]["days"][len(cached_result[0]["days"]) - 1]

            parsed_latest_date = parser.parse(latest_date)
            parsed_latest_date = parsed_latest_date.replace(tzinfo=pytz.timezone(team.timezone))
            _is_present = is_filter_date_present(filter, parsed_latest_date)
        else:
            _is_present = False

        return _is_present

    # Use a condensed filter if a cached result exists in the current timerange
    def adjusted_filter(self, filter: Filter, team: Team) -> Tuple[Filter, Optional[Dict[str, Any]]]:
        cached_result = self.get_cached_result(filter, team)

        new_filter = filter.with_data({"date_from": interval_unit(filter.interval)}) if cached_result else filter

        label_to_payload = {}
        if cached_result:
            for payload in cached_result:
                label_to_payload[f'{payload["label"]}_{payload["action"]["order"]}'] = payload

        return new_filter, label_to_payload

    def merge_results(
        self, result, cached_result: Optional[Dict[str, Any]], entity_order: int, filter: Filter, team: Team
    ):
        if cached_result and filter.display != TRENDS_CUMULATIVE:
            new_res = []

            for payload in result:
                cached_series = cached_result.pop(f'{payload["label"]}_{entity_order}')
                data = cached_series["data"]
                data.pop()
                data.append(payload["data"].pop())
                cached_series["data"] = data
                new_res.append(cached_series)

            return new_res, cached_result
        elif filter.display == TRENDS_CUMULATIVE:
            return self._handle_cumulative(result), {}
        else:
            return result, {}

    def _run_query(self, filter: Filter, team: Team, entity: Entity) -> List[Dict[str, Any]]:
        adjusted_filter, cached_result = self.adjusted_filter(filter, team)
        sql, params, parse_function = self._get_sql_for_entity(adjusted_filter, team, entity)

        result = sync_execute(sql, params)
        result = parse_function(result)
        serialized_data = self._format_serialized(entity, result)
        merged_results, cached_result = self.merge_results(
            serialized_data, cached_result, entity.order or entity.index, filter, team
        )

        if cached_result:
            for value in cached_result.values():
                merged_results.append(value)

        return merged_results

    def _run_query_for_threading(self, result: List, index: int, sql, params):
        result[index] = sync_execute(sql, params)

    def _run_parallel(self, filter: Filter, team: Team) -> List[Dict[str, Any]]:
        result: List[Union[None, List[Dict[str, Any]]]] = [None] * len(filter.entities)
        parse_functions: List[Union[None, Callable]] = [None] * len(filter.entities)
        cached_result = None
        jobs = []

        for entity in filter.entities:
            adjusted_filter, cached_result = self.adjusted_filter(filter, team)
            sql, params, parse_function = self._get_sql_for_entity(adjusted_filter, team, entity)
            parse_functions[entity.index] = parse_function
            thread = threading.Thread(target=self._run_query_for_threading, args=(result, entity.index, sql, params),)
            jobs.append(thread)

        # Start the threads (i.e. calculate the random number lists)
        for j in jobs:
            j.start()

        # Ensure all of the threads have finished
        for j in jobs:
            j.join()

        # Parse results for each thread
        for entity in filter.entities:
            serialized_data = cast(List[Callable], parse_functions)[entity.index](result[entity.index])
            serialized_data = self._format_serialized(entity, serialized_data)
            merged_results, cached_result = self.merge_results(
                serialized_data, cached_result, entity.order or entity.index, filter, team
            )
            result[entity.index] = merged_results

        # flatten results
        flat_results: List[Dict[str, Any]] = []
        for item in result:
            for flat in cast(List[Dict[str, Any]], item):
                flat_results.append(flat)

            if cached_result:
                for value in cached_result.values():
                    flat_results.append(value)

        return flat_results

    def run(self, filter: Filter, team: Team, *args, **kwargs) -> List[Dict[str, Any]]:
        actions = Action.objects.filter(team_id=team.pk).order_by("-id")
        if len(filter.actions) > 0:
            actions = Action.objects.filter(pk__in=[entity.id for entity in filter.actions], team_id=team.pk)
        actions = actions.prefetch_related(Prefetch("steps", queryset=ActionStep.objects.order_by("id")))

        if filter.formula:
            return handle_compare(filter, self._run_formula_query, team)

        for entity in filter.entities:
            if entity.type == TREND_FILTER_TYPE_ACTIONS:
                try:
                    entity.name = actions.get(id=entity.id).name
                except Action.DoesNotExist:
                    return []

        if len(filter.entities) == 1 or filter.compare:
            result = []
            for entity in filter.entities:
                result.extend(handle_compare(filter, self._run_query, team, entity=entity))
        else:
            result = self._run_parallel(filter, team)

        return result

    def _format_serialized(self, entity: Entity, result: List[Dict[str, Any]]):
        serialized_data = []

        serialized: Dict[str, Any] = {
            "action": entity.to_dict(),
            "label": entity.name,
            "count": 0,
            "data": [],
            "labels": [],
            "days": [],
        }

        for queried_metric in result:
            serialized_copy = copy.deepcopy(serialized)
            serialized_copy.update(queried_metric)
            serialized_data.append(serialized_copy)

        return serialized_data

    def _handle_cumulative(self, entity_metrics: List) -> List[Dict[str, Any]]:
        for metrics in entity_metrics:
            metrics.update(data=list(accumulate(metrics["data"])))
        return entity_metrics


def is_filter_date_present(filter: Filter, latest_cached_datetime: datetime) -> bool:
    diff = filter.date_to - latest_cached_datetime

    if filter.interval == "hour":
        return diff < timedelta(hours=1)
    elif filter.interval == "day":
        return diff < timedelta(days=1)
    elif filter.interval == "week":
        return diff < timedelta(weeks=1)
    elif filter.interval == "month":
        return diff < timedelta(days=30)


def interval_unit(interval: str) -> str:
    if interval == "hour":
        return "-1hr"
    if interval == "day":
        return "-1d"
    elif interval == "week":
        return "-1w"
    elif interval == "month":
        return "-1m"
    else:
        raise ValueError("Invalid interval")
