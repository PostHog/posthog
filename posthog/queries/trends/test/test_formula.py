from typing import Dict, Optional

from freezegun.api import freeze_time

from posthog.constants import TRENDS_CUMULATIVE, TRENDS_PIE
from posthog.models import Cohort, Person
from posthog.models.filters.filter import Filter
from posthog.queries.trends.trends import Trends
from posthog.test.base import APIBaseTest, ClickhouseTestMixin, _create_event, snapshot_clickhouse_queries


class TestFormula(ClickhouseTestMixin, APIBaseTest):
    CLASS_DATA_LEVEL_SETUP = False

    def setUp(self):
        super().setUp()

        Person.objects.create(
            team_id=self.team.pk, distinct_ids=["blabla", "anonymous_id"], properties={"$some_prop": "some_val"}
        )
        with freeze_time("2020-01-02T13:01:01Z"):
            _create_event(
                team=self.team,
                event="session start",
                distinct_id="blabla",
                properties={
                    "session duration": 200,
                    "location": "Paris",
                    "$current_url": "http://example.org",
                    "$session_id": "1",
                },
            )
            _create_event(
                team=self.team,
                event="session start",
                distinct_id="blabla",
                properties={"session duration": 300, "location": "Paris", "$session_id": "1",},
            )
            _create_event(
                team=self.team,
                event="session start",
                distinct_id="blabla",
                properties={"session duration": 400, "location": "London", "$session_id": "1",},
            )
        with freeze_time("2020-01-03T13:01:01Z"):
            _create_event(
                team=self.team,
                event="session start",
                distinct_id="blabla",
                properties={"session duration": 400, "location": "London", "$session_id": "1",},
            )
        with freeze_time("2020-01-03T13:04:01Z"):
            _create_event(
                team=self.team,
                event="session start",
                distinct_id="blabla",
                properties={"session duration": 500, "location": "London", "$session_id": "1",},
            )
            _create_event(
                team=self.team,
                event="session end",
                distinct_id="blabla",
                properties={"session duration": 500, "location": "London", "$session_id": "1",},
            )

    def _run(self, extra: Dict = {}, run_at: Optional[str] = None):
        with freeze_time(run_at or "2020-01-04T13:01:01Z"):
            action_response = Trends().run(
                Filter(
                    data={
                        "events": [
                            {"id": "session start", "math": "sum", "math_property": "session duration"},
                            {"id": "session start", "math": "avg", "math_property": "session duration"},
                        ],
                        "formula": "A + B",
                        **extra,
                    }
                ),
                self.team,
            )
        return action_response

    def test_hour_interval(self):
        data = self._run({"date_from": "-1d", "interval": "hour"}, run_at="2020-01-03T13:05:01Z")[0]["data"]
        self.assertEqual(
            data,
            [
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                1200.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                0.0,
                1350.0,
            ],
        )

    def test_day_interval(self):
        data = self._run({"date_from": "-3d"}, run_at="2020-01-03T13:05:01Z")[0]["data"]
        self.assertEqual(data, [0.0, 0.0, 1200.0, 1350.0])

    def test_week_interval(self):
        data = self._run({"date_from": "-2w", "interval": "week"}, run_at="2020-01-03T13:05:01Z")[0]["data"]
        self.assertEqual(data, [0.0, 0.0, 2160.0])

    def test_month_interval(self):
        data = self._run({"date_from": "-2m", "interval": "month"}, run_at="2020-01-03T13:05:01Z")[0]["data"]
        self.assertEqual(data, [0.0, 0.0, 2160.0])

    def test_formula(self):
        self.assertEqual(self._run({"formula": "A - B"})[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 600.0, 450.0, 0.0])
        self.assertEqual(self._run({"formula": "A * B"})[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 270000.0, 405000.0, 0.0])
        self.assertEqual(self._run({"formula": "A / B"})[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 3.0, 2.0, 0.0])
        self.assertEqual(self._run({"formula": "(A/3600)/B"})[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
        self.assertEqual(self._run({"formula": "(A/3600)/B"})[0]["count"], 0)

        self.assertEqual(self._run({"formula": "A/0"})[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])
        self.assertEqual(self._run({"formula": "A/0"})[0]["count"], 0)

    @snapshot_clickhouse_queries
    def test_formula_with_unique_sessions(self):
        with freeze_time("2020-01-04T13:01:01Z"):
            action_response = Trends().run(
                Filter(
                    data={
                        "events": [
                            {"id": "session start", "math": "unique_session"},
                            {"id": "session start", "math": "dau"},
                        ],
                        "formula": "A / B",
                    }
                ),
                self.team,
            )
            self.assertEqual(action_response[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.0])

    def test_breakdown(self):
        response = self._run({"formula": "A - B", "breakdown": "location"})
        self.assertEqual(response[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 450.0, 0.0])
        self.assertEqual(response[0]["label"], "London")
        self.assertEqual(response[1]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 250.0, 0.0, 0.0])
        self.assertEqual(response[1]["label"], "Paris")

    def test_breakdown_counts_of_different_events_one_without_events(self):
        with freeze_time("2020-01-04T13:01:01Z"):
            response = Trends().run(
                Filter(
                    data={
                        "insight": "TRENDS",
                        "display": "ActionsLineGraph",
                        "formula": "B / A",
                        "breakdown": "location",
                        "breakdown_type": "event",
                        "events": [
                            {"id": "session start", "name": "session start", "type": "events", "order": 0},
                            {"id": "session error", "name": "session error", "type": "events", "order": 1},
                        ],
                    }
                ),
                self.team,
            )
        self.assertEqual(
            response,
            [
                {
                    "data": [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                    "count": 0.0,
                    "labels": [
                        "28-Dec-2019",
                        "29-Dec-2019",
                        "30-Dec-2019",
                        "31-Dec-2019",
                        "1-Jan-2020",
                        "2-Jan-2020",
                        "3-Jan-2020",
                        "4-Jan-2020",
                    ],
                    "days": [
                        "2019-12-28",
                        "2019-12-29",
                        "2019-12-30",
                        "2019-12-31",
                        "2020-01-01",
                        "2020-01-02",
                        "2020-01-03",
                        "2020-01-04",
                    ],
                    "label": "London",
                },
                {
                    "data": [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                    "count": 0.0,
                    "labels": [
                        "28-Dec-2019",
                        "29-Dec-2019",
                        "30-Dec-2019",
                        "31-Dec-2019",
                        "1-Jan-2020",
                        "2-Jan-2020",
                        "3-Jan-2020",
                        "4-Jan-2020",
                    ],
                    "days": [
                        "2019-12-28",
                        "2019-12-29",
                        "2019-12-30",
                        "2019-12-31",
                        "2020-01-01",
                        "2020-01-02",
                        "2020-01-03",
                        "2020-01-04",
                    ],
                    "label": "Paris",
                },
            ],
        )

    def test_breakdown_cohort(self):
        cohort = Cohort.objects.create(
            team=self.team,
            name="cohort1",
            groups=[{"properties": [{"key": "$some_prop", "value": "some_val", "type": "person"}]}],
        )
        response = self._run({"breakdown": ["all", cohort.pk], "breakdown_type": "cohort"})
        self.assertEqual(response[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 1200.0, 1350.0, 0.0])
        self.assertEqual(response[0]["label"], "all users")
        self.assertEqual(response[1]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 1200.0, 1350.0, 0.0])
        self.assertEqual(response[1]["label"], "cohort1")

    def test_breakdown_mismatching_sizes(self):
        response = self._run(
            {"events": [{"id": "session start"}, {"id": "session end"},], "breakdown": "location", "formula": "A + B",}
        )

        self.assertEqual(response[0]["label"], "London")
        self.assertEqual(response[0]["data"], [0, 0, 0, 0, 0, 1, 3, 0])
        self.assertEqual(response[1]["label"], "Paris")
        self.assertEqual(response[1]["data"], [0, 0, 0, 0, 0, 2, 0, 0])

    def test_global_properties(self):
        self.assertEqual(
            self._run({"properties": [{"key": "$current_url", "value": "http://example.org"}]})[0]["data"],
            [0.0, 0.0, 0.0, 0.0, 0.0, 400.0, 0.0, 0.0],
        )

    def test_properties_with_escape_params(self):
        # regression test
        self.assertEqual(
            self._run(
                {
                    "properties": [
                        {
                            "key": "$current_url",
                            "value": "http://localhost:8000/insights?insight=TRENDS&interval=day&display=ActionsLineGraph&actions=%5B%5D&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%7D%2C%7B%22id%22%3A%22%24pageview%2",
                        }
                    ]
                }
            )[0]["data"],
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        )

    def test_event_properties(self):
        self.assertEqual(
            self._run(
                {
                    "events": [
                        {
                            "id": "session start",
                            "math": "sum",
                            "math_property": "session duration",
                            "properties": [{"key": "$current_url", "value": "http://example.org"}],
                        },
                        {"id": "session start", "math": "avg", "math_property": "session duration"},
                    ]
                }
            )[0]["data"],
            [0.0, 0.0, 0.0, 0.0, 0.0, 500.0, 450.0, 0.0],
        )

    def test_compare(self):
        response = self._run({"date_from": "-1dStart", "compare": True})
        self.assertEqual(response[0]["data"], [1350.0, 0.0])
        self.assertEqual(response[1]["data"], [1200.0, 1350.0])

    def test_pie(self):
        self.assertEqual(self._run({"display": TRENDS_PIE})[0]["aggregated_value"], 2160.0)

    def test_cumulative(self):
        self.assertEqual(
            self._run({"display": TRENDS_CUMULATIVE})[0]["data"], [0.0, 0.0, 0.0, 0.0, 0.0, 1200.0, 2550.0, 2550.0]
        )

    def test_multiple_events(self):
        # regression test
        self.assertEqual(
            self._run(
                {
                    "events": [
                        {"id": "session start", "math": "sum", "math_property": "session duration"},
                        {"id": "session start", "math": "avg", "math_property": "session duration"},
                        {"id": "session start", "math": "avg", "math_property": "session duration"},
                    ]
                }
            )[0]["data"],
            [0.0, 0.0, 0.0, 0.0, 0.0, 1200.0, 1350.0, 0.0],
        )
