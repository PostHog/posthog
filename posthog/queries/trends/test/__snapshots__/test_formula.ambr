# name: TestFormula.test_formula_with_unique_sessions
  '
  SELECT sub_A.date,
         arrayMap((A, B) -> A / B, arrayResize(sub_A.data, max_length, 0), arrayResize(sub_B.data, max_length, 0)) ,
         arrayMax([length(sub_A.data), length(sub_B.data)]) as max_length
  FROM
    (SELECT groupArray(day_start) as date,
            groupArray(count) as data
     FROM
       (SELECT SUM(total) AS count,
               day_start
        from
          (SELECT toUInt16(0) AS total,
                  toStartOfDay(toDateTime('2020-01-04 23:59:59') - toIntervalDay(number), 'UTC') AS day_start
           FROM numbers(dateDiff('day', toStartOfDay(toDateTime('2019-12-28 00:00:00'), 'UTC'), toDateTime('2020-01-04 23:59:59'), 'UTC'))
           UNION ALL SELECT toUInt16(0) AS total,
                            toStartOfDay(toDateTime('2019-12-28 00:00:00'), 'UTC')
           UNION ALL SELECT count(DISTINCT $session_id) as data,
                            toStartOfDay(toDateTime(timestamp), 'UTC') as date
           FROM
             (SELECT e.timestamp as timestamp,
                     e."$session_id" as "$session_id"
              FROM events e
              WHERE team_id = 2
                AND event = 'session start'
                AND timestamp >= toStartOfDay(toDateTime('2019-12-28 00:00:00'), 'UTC')
                AND timestamp <= toDateTime('2020-01-04 23:59:59') )
           GROUP BY date)
        group by day_start
        order by day_start SETTINGS allow_experimental_window_functions = 1) SETTINGS timeout_before_checking_execution_speed = 60) as sub_A
  CROSS JOIN
    (SELECT groupArray(day_start) as date,
            groupArray(count) as data
     FROM
       (SELECT SUM(total) AS count,
               day_start
        from
          (SELECT toUInt16(0) AS total,
                  toStartOfDay(toDateTime('2020-01-04 23:59:59') - toIntervalDay(number), 'UTC') AS day_start
           FROM numbers(dateDiff('day', toStartOfDay(toDateTime('2019-12-28 00:00:00'), 'UTC'), toDateTime('2020-01-04 23:59:59'), 'UTC'))
           UNION ALL SELECT toUInt16(0) AS total,
                            toStartOfDay(toDateTime('2019-12-28 00:00:00'), 'UTC')
           UNION ALL SELECT count(DISTINCT person_id) as data,
                            toStartOfDay(toDateTime(timestamp), 'UTC') as date
           FROM
             (SELECT e.timestamp as timestamp,
                     e."$session_id" as "$session_id",
                     pdi.person_id as person_id
              FROM events e
              INNER JOIN
                (SELECT distinct_id,
                        argMax(person_id, version) as person_id
                 FROM person_distinct_id2
                 WHERE team_id = 2
                 GROUP BY distinct_id
                 HAVING argMax(is_deleted, version) = 0) AS pdi ON e.distinct_id = pdi.distinct_id
              WHERE team_id = 2
                AND event = 'session start'
                AND timestamp >= toStartOfDay(toDateTime('2019-12-28 00:00:00'), 'UTC')
                AND timestamp <= toDateTime('2020-01-04 23:59:59') )
           GROUP BY date)
        group by day_start
        order by day_start SETTINGS allow_experimental_window_functions = 1) SETTINGS timeout_before_checking_execution_speed = 60) as sub_B
  '
---
